obspy/clients/fdsn/mass_downloader/domain.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Domain definitions for the the download helpers.
+
+Subclass the :class:`~obspy.clients.fdsn.mass_downloader.domain.Domain` class
+to define custom and potentially more complex domains. See its documentation
+for an example.
+
+:copyright:
+    Lion Krischer (krischer@geophysik.uni-muenchen.de), 2014-2015
+:license:
+    GNU Lesser General Public License, Version 3
+    (http://www.gnu.org/copyleft/lesser.html)
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+from future.utils import with_metaclass
+
+from abc import ABCMeta, abstractmethod
+
+
+class Domain(with_metaclass(ABCMeta)):
+    """
+    Abstract base class defining a domain - subclass it to define a new domain.
+
+    Each subclass has to implement the :meth:`~.get_query_parameters`
+    method and optionally the :meth:`~.is_in_domain` method which enables
+    the construction of arbitrarily complex domains. The
+    :meth:`~.get_query_parameters` method must return the query parameters to
+    download as much data as required. The :meth:`~.is_in_domain` can later be
+    used to refine the domain after the data has been downloaded.
+
+
+    It can be thought of as a boolean operation - first the rough domain is
+    specified including all the possible points, then some points are
+    removed again. This is illustrated with an example domain representing
+    Germany utilizing the external packages
+    `shapely <http://toblerity.org/shapely/>`_ and
+    `fiona <http://toblerity.org/fiona/>`_. Shapefiles can be found online
+    on many websites, for example on
+    `http://www.gadm.org/ <http://www.gadm.org/>`_. The example works by
+    first extracting the bounds of the country to formulate the FDSN query
+    and then removing points outside of the exact shape.
+
+
+    .. code-block:: python
+
+        import fiona
+        import shapely.geometry
+
+        from obspy.clients.fdsn.mass_downloader import Domain
+
+
+        class Germany(Domain):
```

Comment:
I was just thinking some kind of shapely domain might be useful, but I guess the example's right here.